### PR TITLE
Update the structure of api_callable.

### PR DESCRIPTION
**Pull request by jmuk**
Now this allows getting call options as an optional parameter
and changes its behavior.

In a part of #104 effort.

### CODE REVIEW

File: google/gax/__init__.py [hunk 1]
```diff
@@ -80,11 +74,21 @@ def __init__(self, timeout=30, retry=None, page_descriptor=None,
...
         self.page_descriptor = page_descriptor
-        self.flatten_pages = flatten_pages
         self.page_token = page_token
         self.bundler = bundler
         self.bundle_descriptor = bundle_descriptor
 
+    @property
+    def flatten_pages(self):
+        """
+        A boolean property whether a page streamed response should make
```
Comment (geigerj) [inline google/gax/__init__.py]: "property whether" --> "property indicating whether"

File: google/gax/__init__.py [hunk 2]
```diff
@@ -80,11 +74,21 @@ def __init__(self, timeout=30, retry=None, page_descriptor=None,
...
+        the page structure transparent to the user by flattening the
+        repeated field in the returned generator.
```
Comment (geigerj) [inline google/gax/__init__.py]: s/generator/iterator

File: google/gax/__init__.py [hunk 3]
```diff
@@ -103,8 +107,9 @@ def merge(self, options):
         if not options:
             return CallSettings(
                 timeout=self.timeout, retry=self.retry,
-                page_descriptor=self.page_descriptor, bundler=self.bundler,
-                bundle_descriptor=self.bundle_descriptor)
+                page_descriptor=self.page_descriptor,
+                page_token=self.page_token,
+                bundler=self.bundler, bundle_descriptor=self.bundle_descriptor)
```
Comment (geigerj) [inline google/gax/__init__.py]: Good catch!

File: google/gax/api_callable.py [hunk 4]
```diff
@@ -41,59 +41,63 @@
 _MILLIS_PER_SECOND = 1000
 
 
-def _add_timeout_arg(a_func, timeout):
-    """Updates a_func so that it gets called with the timeout as its final arg.
+class _RetryableWrapper(object):
```
Comment (geigerj) [inline google/gax/api_callable.py]: @tbetbetbe made a past decision to move away from using callable objects to using our current functional/"decorator-like" style. What is the advantage in using an object here?

File: google/gax/api_callable.py [hunk 5]
```diff
@@ -41,59 +41,63 @@
...
 
-    Returns:
-      callable: the original callable updated to the timeout arg
-    """
+        Args:
+          req (object): the request object.
+          options (gax.CallOptions): the optional value to customize the
+            calling behavior.
+        """
+        settings = self._settings.merge(options)
```
Comment (geigerj) [inline google/gax/api_callable.py]: It seems uninuitive that the settings/options merge is happening inside an object called _RetryableWrapper, which I would expect to deal only with retrying. I would rather this happen in `create_api_call`.

Comment (jmuk) [inline google/gax/api_callable.py]: `create_api_call` is not the case because, as is discussed in #104, `create_api_call` is called in the constructor like
```
self._create_shelf = gax.create_api_call(stub.create_shelf, defaults['create_shelf'])
```
and then the call options is passed in the individual method invocation, like:
```
def create_shelf(self, options):
   self._create_shelf(req, options, ...)
```

So the result of `create_api_call` should receive a CallOptions instance, and handling the overrides -- that is the role of merge method.

Comment (geigerj) [inline google/gax/api_callable.py]: What about adopting the same structure for `create_api_call` as for the rest of the decorator-style functions?

```
def create_api_call(...)
  def inner(...)
    settings.merge(options)
    ...
```

?

File: google/gax/api_callable.py (hunk 4 above)
Comment (jmuk) [inline google/gax/api_callable.py]: I could rewrite as decorator-style, but I think the code structure is mostly same and using callable class is much more straightforward in this case.

The point is that the decoration (the behavior) may change upon the invocation because it needs to change either of _add_timeout_arg or _retry, and they share some values like a_func and base settings. This is naturally a class, although we could implement it as internal functions.

File: google/gax/__init__.py (hunk 1 above)
Comment (jmuk) [inline google/gax/__init__.py]: done

File: google/gax/__init__.py (hunk 2 above)
Comment (jmuk) [inline google/gax/__init__.py]: done

File: google/gax/api_callable.py (hunk 4 above)
Comment (geigerj) [inline google/gax/api_callable.py]: I don't feel strongly about this, but I don't think this is particularly unnaturally expressed as a function either -- it seems to me that the only "object-y" thing about this is storing the read-only attribute `settings`, which works equally well in a closure.

I'm fine with it either way, although I would rather `settings` belonged to a closure (or callable object) in `create_api_call` than in the retrying part of the code.

**Comment by geigerj** [comment]
Note: googleapis/toolkit#214 also changes the handling of `metadata`, which requires an update to `CallSettings` and `api_callable`.

**Comment by jmuk** [comment]
I reviewed #104 discussion but unclear why metadata goes into construct_settings. It's actually not discussed well.

But I don't think that's the key part of the redesign, and therefore I think that's no problem to be in a part of method invocation, i.e.
```
self._create_shelf(request, options, metadata=self._headers)
```

And this structure would work well with this PR.

File: google/gax/api_callable.py (hunk 5 above)
Comment (jmuk) [inline google/gax/api_callable.py]: I agree that settings.merge() will be in a single place, and restructured a bit. Please take another look. I am not sure this is the right direction though -- adding more and more inner functions of inner functions.

To be clear, this still determine _page_streamable/_bundleable/else when create_api_call() is invoked, for multiple reasons:
- this structure is actually determined when create_api_call() is called, so this is better and more understandable
- a test case actually require to raise an error for inconsistent bundling/page-streaming for create_api_call(), and I think that's the right thing to keep it.

**Comment by geigerj** [comment]
regarding metadata -- this was in Garrett's original mockup on the pubsub codegen, and is implicitly represented in #104 by the simplified method invocation in the second code block. If you feel that this should not be addressed in this PR, I can do it in a follow-up.

File: google/gax/api_callable.py [hunk 6]
```diff
@@ -141,49 +142,48 @@ def _bundleable(a_func, desc, bundler):
...
         the_id = bundling.compute_bundle_id(
             request, desc.request_discriminator_fields)
         return bundler.schedule(a_func, the_id, desc, request, kwargs)
 
     return inner
 
 
-def _page_streamable(a_func, page_descriptor, page_token=None,
-                     flatten_pages=True):
+def _page_streamable(a_func, settings, page_descriptor):
```
Comment (geigerj) [inline google/gax/api_callable.py]: The `page_descriptor` parameter is redundant with the `settings` parameter -- let's just use `settings`

Comment (geigerj) [inline google/gax/api_callable.py]: whoops, was on the wrong commit

File: google/gax/__init__.py [hunk 7]
```diff
@@ -429,3 +431,33 @@ def next(self):
...
+        self._current = None
+        self._index = -1
+
+    def __iter__(self):
+        return self
+
+    def next(self):
+        """Retrieves the next resource."""
+        # pylint: disable=next-method-called
+        while not self._current:
```
Comment (geigerj) [inline google/gax/__init__.py]: Why the while loop? That is, when will a PageIterator contain a falsy value?

Comment (jmuk) [inline google/gax/__init__.py]: PageIterator could return an empty array but have a next page token, in that case this should skip the empty array and fetch the next page.

**Comment by jmuk** [comment]
Regarding metadata -- rethinking about that, I'm okay with merging into construct_settings.
But I'm thinking `CallSettings` might not be a good name now -- it keeps the option parameters (like backoff settings) and the information of the API-call such as page descriptions, bundle descriptions, and now metadata (headers) -- those are properties of the method itself rather than invocations (= call).
This could be `MethodSettings`?

**Comment by geigerj** [comment]
SGTM regarding `CallSettings` vs `MethodSettings`. I think there has always been some ambiguity from the naming of `CallOptions` and `CallSettings` about the difference -- hopefully this name change will make it clearer.

**Comment by jmuk** [comment]
Updated about the metadata -- not renamed the class yet. I think it can be done in another PR.

Also, I think GAX should not be strongly tied to gRPC itself -- it should be encapsulated into grpc.py file only as far as possible, and therefore, I've added 'kwargs' argument instead.

File: google/gax/__init__.py [hunk 8]
```diff
@@ -134,7 +139,8 @@ def merge(self, options):
             return CallSettings(
                 timeout=timeout, retry=retry,
                 page_descriptor=self.page_descriptor, page_token=page_token,
-                bundler=bundler, bundle_descriptor=self.bundle_descriptor)
+                bundler=bundler, bundle_descriptor=self.bundle_descriptor,
+                kwargs=self.kwargs)
```
Comment (geigerj) [inline google/gax/__init__.py]: Let's a actually do

```
if options.kwargs == OPTION_INHERIT:
  kwargs = self.kwargs
else:
  kwargs = self.kwargs.copy()
  kwargs.update(options.kwargs)
```

I can't tell where `options.kwargs` is currently being used -- maybe we are just ignoring it at the moment?

File: google/gax/api_callable.py [hunk 9]
```diff
@@ -88,7 +88,7 @@ def _retryable(a_func, retry):
     total_timeout = (retry.backoff_settings.total_timeout_millis /
                      _MILLIS_PER_SECOND)
 
-    def inner(*args, **kwargs):
```
Comment (geigerj) [inline google/gax/api_callable.py]: This is a good cleanup -- previously, both CallOptions and the inner call took `kwargs`, which was redundant. This is better.

Comment (geigerj) [inline google/gax/api_callable.py]: Optional: we should ensure that the "decorator"/transformation functions all look like this, i.e., the final call's `kwargs` comes only from Call{Settings,Options} and not from actual function parameters. This can also be done in a follow-up PR.

**Comment by geigerj** [comment]
Just one concern about the CallSettings `merge` method. Otherwise looks good

File: google/gax/__init__.py (hunk 8 above)
Comment (jmuk) [inline google/gax/__init__.py]: added -- no such options.kwargs exist currently, so I've added it.

File: google/gax/api_callable.py (hunk 9 above)
Comment (jmuk) [inline google/gax/api_callable.py]: agreed, let me do it in another PR

**Comment by geigerj** [comment]
LGTM.

I will update googleapis/toolkit#214 to put metadata into kwargs.
